### ISSUE TITLE
Release binary crashes on MacOS 11.0.1

### ISSUE THREAD

**Issue by bchmnn** [onivim/oni2#2813]
### Versions:
* MacOS: 11.0.1
* node: v15.0.3
* npm: 7.0.14
* esy: 0.6.7
* onivim/oni2: commit 533d1035
* fish shell: 3.1.2
* alacritty: 0.6.0 (04cbf76)

### Bug:
The release binary of Onivim2 crashes immediatelly after execution. The window pops up for just a millisecond.
Interesting is that executing Onivim2 by calling `esy run` works just fine.

### Log:
```bash
$ esy '@release' run -f --checkhealth
dyld: Library not loaded: @rpath/Sparkle.framework/Versions/A/Sparkle
  Referenced from: /Users/bchmnn/Development/oni2/_esy/release/store/i/oni2-fb471e05/bin/Oni2_editor
  Reason: image not found
```
```bash
$ _esy/release/install/bin/Oni2 -f --trace --debug
dyld: Library not loaded: @rpath/Sparkle.framework/Versions/A/Sparkle
  Referenced from: /Users/bchmnn/Development/oni2/./_esy/release/install/bin/Oni2_editor
  Reason: image not found
```
```bash
$ _release/Onivim2.app/Contents/MacOS/Oni2 -f --trace --debug
[DEBUG] [ 0.000s] Oni2.Exception : Recording backtraces
[INFO]  [ 0.000s] Oni2_editor : Starting Onivim 2.0.5.1-nightly (01afc541)
[DEBUG] [ 0.000s] Oni2_editor : Calling App.start
[DEBUG] [ 0.053s] Oni2_editor : Init
[INFO]  [ 0.068s] Oni2_editor : Startup: Trying to change folder to: /Users/bchmnn/Development/oni2
[INFO]  [ 0.068s] Oni2_editor :  - Have read permission
[INFO]  [ 0.069s] Oni2_editor : - Ran chdir
[INFO]  [ 0.069s] Oni2_editor : Successfully changed working directory to: /Users/bchmnn/Development/oni2
[INFO]  [ 0.069s] Oni2_editor : Unsanitized x value: 25
[INFO]  [ 0.069s] Oni2_editor : Unsanitized x value: 0
[INFO]  [ 0.069s] Oni2_editor : Sanitized values from persistence - x: Absolute(0) y: Absolute(25) width: 1440 height: 875
[DEBUG] [ 0.069s] Revery.Core.Window : Starting window creation...
[INFO]  [ 0.069s] Revery.Core.Window : Display count: 1
[INFO]  [ 0.069s] Revery.Core.Window : Display 0:Color LCD - pixelFormat: SDL_PIXELFORMAT_ARGB8888 width: 1440 height: 900 refreshRate: 60 Bounds: x: 0 y: 0 width: 1440 height: 900
[INFO]  [ 0.069s] Revery.Core.Window : Creating window Oni2 width: 1440 height: 875
[INFO]  [ 0.100s] Revery.Core.Window : Window created successfully.
[DEBUG] [ 0.100s] Revery.Core.Window : - Id: 1
[DEBUG] [ 0.101s] Revery.Core.Window : - PixelFormat: SDL_PIXELFORMAT_ARGB8888
[DEBUG] [ 0.101s] Revery.Core.Window : Setting window context
[DEBUG] [ 0.122s] Revery.Core.Window : GL setup. Checking GL version...
[DEBUG] [ 0.122s] Revery.Core.Window : Checking GL vendor...
[DEBUG] [ 0.122s] Revery.Core.Window : Checking GL shading language version...
[INFO]  [ 0.122s] Revery.Core.Window : OpenGL hardware info:
[INFO]  [ 0.122s] Revery.Core.Window :   renderer: Intel(R) Iris(TM) Plus Graphics 655
[INFO]  [ 0.122s] Revery.Core.Window :   version: 4.1 INTEL-16.0.49
[INFO]  [ 0.122s] Revery.Core.Window :   vendor: Intel Inc.
[INFO]  [ 0.122s] Revery.Core.Window :   shadingLanguageVersion: 4.10
[DEBUG] [ 0.122s] Revery.Core.Window : Loading icon from: /Users/bchmnn/Development/oni2/_release/Onivim2.app/Contents/MacOS/logo.png
[ERROR] [ 0.122s] Revery.Core.Window : Error loading icon: can't fopen
[DEBUG] [ 0.122s] Revery.Core.Window : Getting window metrics
[DEBUG] [ 0.122s] Revery.Core.Window : Metrics: { scaledSize = { width = 1440; height = 847 };
  unscaledSize = { width = 1440; height = 847 };
  framebufferSize = { width = 2880; height = 1694 }; devicePixelRatio = 2.;
  scaleFactor = 1.; zoom = 1.; isDirty = false }
[INFO]  [ 0.123s] Revery.Core.Window : Using vsync: Immediate
[DEBUG] [ 0.201s] Oni2_editor : Initializing setup.
[DEBUG] [ 0.201s] Oni2.Core.Setup : Looking for setup configuration at: /Users/bchmnn/Development/oni2/_release/Onivim2.app/Contents/MacOS/setup.json
[WARN]  [ 0.222s] Revery.FontCache : Error loading typeface (skia)
[ERROR] [ 0.225s] Oni2.Exception : Exception (Invalid_argument "result is Error _"):
Raised at file "stdlib.ml", line 30, characters 20-45
Called from file "src/Core/Utility/OptionEx.re" (inlined), line 65, characters 12-15
Called from file "src/Core/FontMeasurementCache.re", line 20, characters 4-352
Called from file "src/Core/Font.re", line 33, characters 6-121
Called from file "src/bin_editor/Oni2_editor.re", line 247, characters 14-36
Called from file "src/Core/App.re", line 199, characters 17-34
Called from file "packages/reason-sdl2/src/sdl2.re", line 269, characters 17-59
Called from file "packages/reason-sdl2/src/sdl2.re", line 269, characters 17-59
Called from file "src/bin_editor/Oni2_editor.re", line 481, characters 11-26
```

**Comment by bryphe** [comment]
Unfortunately I haven't tested the build yet on Big Sur (the release builds we produce are built from OSX 10.14).

I've read that there are some different linking requirements for Big Sur / `dyld` - in particular, we may need to codesign binaries even on development machines (currently - we only codesign for release buidls).

**Comment by zbaylin** [comment]
This is actually to be expected, since macOS has for a while forbidden dynamic linking of frameworks outside of app bundles.  @bchmnn try running `esy @release create --codesign` and running the produced bundle and it should work. The reason we keep the `run` command for the `release` sandbox is because esy by default doesn't copy over the files into the correct places that we need for release creation until you try to run something.

**Comment by bchmnn** [comment]
@zbaylin Unfortunately, his still does not work..
Replaced irrelevant information with ".........................." in log:
```
$ esy '@release' create --codesign
..........................
[shell] dylibbundler -b -x "/Users/bchmnn/Development/oni2/_release/Onivim2.app/Contents/MacOS/Oni2_editor" -d "/Users/bchmnn/Development/oni2/_release/Onivim2.app/Contents/Frameworks" -p "@executable_path/../Frameworks/" -cd

/!\ WARNING : Cannot resolve path '/usr/lib/libiconv.2.dylib'

/!\ WARNING : Library libiconv.2.dylib has an incomplete name (location unknown)
libiconv.2.dylib does not exist. Try again
libiconv.2.dylib does not exist. Try again
..........................
node:child_process:675
    throw err;
    ^

<ref *1> Error: spawnSync /bin/sh ENOBUFS
    at Object.spawnSync (node:internal/child_process:1074:20)
    at spawnSync (node:child_process:622:24)
    at Object.execSync (node:child_process:667:15)
    at shell (/Users/bchmnn/Development/oni2/scripts/release.js:51:20)
    at Object.<anonymous> (/Users/bchmnn/Development/oni2/scripts/release.js:240:5)
    at Module._compile (node:internal/modules/cjs/loader:1102:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1131:10)
    at Module.load (node:internal/modules/cjs/loader:967:32)
    at Function.Module._load (/Users/bchmnn/Development/oni2/_esy/release/pnp.js:3604:14)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:76:12) {
  errno: -55,
  code: 'ENOBUFS',
  syscall: 'spawnSync /bin/sh',
  path: '/bin/sh',
  spawnargs: [
    '-c',
    'dylibbundler -b -x "/Users/bchmnn/Development/oni2/_release/Onivim2.app/Contents/MacOS/Oni2_editor" -d "/Users/bchmnn/Development/oni2/_release/Onivim2.app/Contents/Frameworks" -p "@executable_path/../Frameworks/" -cd'
  ],
  error: [Circular *1],
  status: null,
  signal: 'SIGTERM',
  output: [
    null,
    Buffer(719421) [Uint8Array] [
       42,  32,  67, 111, 108, 108, 101,  99, 116, 105, 110, 103,
       32, 100, 101, 112, 101, 110, 100, 101, 110,  99, 105, 101,
      115,  46,  46,  46,  46,  46,  46,  46,  46,  46,  46,  46,
       46,  46,  46,  80, 108, 101,  97, 115, 101,  32, 115, 112,
      101,  99, 105, 102, 121,  32, 110, 111, 119,  32, 116, 104,
      101,  32, 100, 105, 114, 101,  99, 116, 111, 114, 121,  32,
      119, 104, 101, 114, 101,  32, 116, 104, 105, 115,  32, 108,
      105,  98, 114,  97, 114, 121,  32,  99,  97, 110,  32,  98,
      101,  32, 102, 111,
      ... 719321 more items
    ],
    Buffer(329181) [Uint8Array] [
       10,  47,  33,  92,  32,  87,  65,  82,  78,  73,  78,  71,
       32,  58,  32,  67,  97, 110, 110, 111, 116,  32, 114, 101,
      115, 111, 108, 118, 101,  32, 112,  97, 116, 104,  32,  39,
       47, 117, 115, 114,  47, 108, 105,  98,  47, 108, 105,  98,
      105,  99, 111, 110, 118,  46,  50,  46, 100, 121, 108, 105,
       98,  39,  10,  10,  47,  33,  92,  32,  87,  65,  82,  78,
       73,  78,  71,  32,  58,  32,  76, 105,  98, 114,  97, 114,
      121,  32, 108, 105,  98, 105,  99, 111, 110, 118,  46,  50,
       46, 100, 121, 108,
      ... 329081 more items
    ]
  ],
  pid: 75316,
  stdout: Buffer(719421) [Uint8Array] [
     42,  32,  67, 111, 108, 108, 101,  99, 116, 105, 110, 103,
     32, 100, 101, 112, 101, 110, 100, 101, 110,  99, 105, 101,
    115,  46,  46,  46,  46,  46,  46,  46,  46,  46,  46,  46,
     46,  46,  46,  80, 108, 101,  97, 115, 101,  32, 115, 112,
    101,  99, 105, 102, 121,  32, 110, 111, 119,  32, 116, 104,
    101,  32, 100, 105, 114, 101,  99, 116, 111, 114, 121,  32,
    119, 104, 101, 114, 101,  32, 116, 104, 105, 115,  32, 108,
    105,  98, 114,  97, 114, 121,  32,  99,  97, 110,  32,  98,
    101,  32, 102, 111,
    ... 719321 more items
  ],
  stderr: Buffer(329181) [Uint8Array] [
     10,  47,  33,  92,  32,  87,  65,  82,  78,  73,  78,  71,
     32,  58,  32,  67,  97, 110, 110, 111, 116,  32, 114, 101,
    115, 111, 108, 118, 101,  32, 112,  97, 116, 104,  32,  39,
     47, 117, 115, 114,  47, 108, 105,  98,  47, 108, 105,  98,
    105,  99, 111, 110, 118,  46,  50,  46, 100, 121, 108, 105,
     98,  39,  10,  10,  47,  33,  92,  32,  87,  65,  82,  78,
     73,  78,  71,  32,  58,  32,  76, 105,  98, 114,  97, 114,
    121,  32, 108, 105,  98, 105,  99, 111, 110, 118,  46,  50,
     46, 100, 121, 108,
    ... 329081 more items
  ]
}
```

**Comment by bchmnn** [comment]
I also checked `/usr/lib` and it indeed does not contain `libiconv.2.dylib`:
```
$ /bin/ls /usr/lib
charset.alias			libpython2.7.dylib
cron				libstdc++.6.dylib
dsc_extractor.bundle		libstdc++.dylib
dtrace				log
dyld				pam
groff				php
libLeaksAtExit.dylib		pkgconfig
libMTLCapture.dylib		python2.7
libffi-trampolines.dylib	rpcsvc
libgmalloc.dylib		ruby
libhunspell-1.2.0.dylib		sasl2
libhunspell-1.2.dylib		sqlite3
libiodbc.2.dylib		swift
libiodbc.dylib			system
libiodbcinst.2.dylib		updaters
libiodbcinst.dylib		xpc
libobjc-trampolines.dylib	zsh
libpython.dylib
```

**Comment by proful** [comment]
Same error as @bchmnn in Big Sur
`libiconv.2.dylib does not exist. Try again`

**Comment by bchmnn** [comment]
Any news here?

**Comment by brdoney** [comment]
I figured that this would become an issue at some point in the future, so I worked at it for a while and here's what I figured out (through lots of trial and error haha):

1. `esy-macdylibbundler` needs to be updated; a fix came out for Big Sur auriamg/macdylibbundler#54, but a new 
version of the package hasn't been published
2. Using symlinks for the `Oni2` and `Oni2_Editor` doesn't seem to work. Instead of copying the symlinks from esy's `$cur__bin`, maybe they could be explicitly replaced with the binaries that they point to?

As a manual fix, I created a more up-to-date version of `esy-macdylibbundler` locally using auriamg/macdylibbundler@8b7a98b (totally arbitrary, it's just after the Big Sur fixes) as a base and edited esy's resolutions to point to it. Plus, I manually replaced the symlinks for `Oni2` and `Oni2_Editor` in `_release/Onivim2.app/Contents/MacOS/` with their actual binaries after creating the release. I can post more info if needed.

Once the above issues are fixed and the app is built, I noticed the following:
* Codesigning, at least for a local release build, doesn't matter
* `esy '@release' run -f --checkhealth` doesn't work (it results in the Sparkle linking error), but `_release/Onivim2.app/Contents/MacOS/Oni2 -f --checkhealth` works. I'm guessing this has to do with Sparkle being where it's expected when using the app version, but I'm not sure

So **TLDR**, `esy-macdylibbundler` needs to be updated to use a newer version of [auriamg/macdylibbundler](https://github.com/auriamg/macdylibbundler) as its base and the `Oni2` and `Oni_Editor` need to be converted from symlinks to their real binaries.

I can open a PR on the latter part, but @bryphe I think you host `esy-macdylibbundler`. Is it possible to get it updated?

**Comment by brdoney** [comment]
I just found bryphe/esy-macdylibbundler, so I put in a PR to update the files there @bryphe (sorry for another @). Of course, the version there still needs to be bumped, published, and then we need to update the dependency here, but that's all it should take.

**Comment by bryphe** [comment]
Thanks @brdoney for your help with the investigation - _huge_ help! 🎉  

Nice job diving in and figuring out those 2 blockers, and catching that there was an upstream fix for macdylibbundler.

I merged in your PR for `esy-macdylibbundler` and published a new version (`0.4.5001` - bringing it in with #3319) and  your PR #3313 looks great - I just need to test with our internal code-signing / notarization infrastructure - but hopefully #3313 and #3319 together will unblock building on Big Sur 🤞
